File: triad-sample/app/src/test/java/com/nhaarman/triad/sample/editnote/EditExistingNotePresenterTest.java
package com.nhaarman.triad.sample.editnote;

import android.content.Context;
import android.content.res.Resources;
import com.nhaarman.triad.Optional;
import com.nhaarman.triad.Triad;
import com.nhaarman.triad.sample.ActivityComponent;
import com.nhaarman.triad.sample.MemoryNoteRepository;
import com.nhaarman.triad.sample.Note;
import com.nhaarman.triad.sample.NoteCreator;
import com.nhaarman.triad.sample.NoteRepository;
import com.nhaarman.triad.sample.NoteValidator;
import org.junit.Before;
import org.junit.Test;

import static org.mockito.Matchers.any;
import static org.mockito.Mockito.doReturn;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.spy;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

public class EditExistingNotePresenterTest {

  private static final String TITLE = "title";

  private static final String CONTENTS = "contents";

  private EditNotePresenter mEditNotePresenter;

  private EditNoteContainer mEditNoteContainerMock;

  private Note mNote;

  private NoteRepository mNoteRepository;

  private NoteValidator mNoteValidator;

  private Triad mTriad;

  @Before
  public void setUp() {
    mNote = new Note();
    mNote.setTitle(TITLE);
    mNote.setContents(CONTENTS);

    mNoteValidator = spy(new NoteValidator());
    mNoteRepository = spy(new MemoryNoteRepository());

    mTriad = mock(Triad.class);

    mEditNotePresenter = spy(new EditNotePresenter(mNote, mNoteValidator, mock(NoteCreator.class), mNoteRepository, mTriad));
    doReturn(Optional.of(mock(Resources.class))).when(mEditNotePresenter).getResources();

    mEditNoteContainerMock = mock(EditNoteContainer.class);
    when(mEditNoteContainerMock.getContext()).thenReturn(mock(Context.class));
  }

  @Test
  public void onControlGained_setsNoteTitleToEditNoteContainer() {
    /* Given */
    EditNoteContainer containerMock = mock(EditNoteContainer.class);
    ActivityComponent activityComponentMock = mock(ActivityComponent.class);

    /* When */
    mEditNotePresenter.onControlGained(containerMock, activityComponentMock);

    /* Then */
    verify(containerMock).setTitle(TITLE);
  }

  @Test
  public void onControlGained_setsNoteContentsToEditNoteContainer() {
    /* Given */
    EditNoteContainer containerMock = mock(EditNoteContainer.class);
    ActivityComponent activityComponentMock = mock(ActivityComponent.class);

    /* When */
    mEditNotePresenter.onControlGained(containerMock, activityComponentMock);

    /* Then */
    verify(containerMock).setContents(CONTENTS);
  }

  @Test
  public void onSaveNoteClicked_validatesTitle() {
    /* Given */
    mEditNotePresenter.setContainer(mEditNoteContainerMock);

    when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
    when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mNoteValidator).validateTitle(TITLE);
  }

  @Test
  public void onSaveNoteClicked_validatesContents() {
    /* Given */
    mEditNotePresenter.setContainer(mEditNoteContainerMock);

    when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
    when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mNoteValidator).validateContents(CONTENTS);
  }

  @Test
  public void onSaveNoteClickedWithEmptyTitle_showsErrorMessage() {
    /* Given */
    mEditNotePresenter.setContainer(mEditNoteContainerMock);

    when(mEditNoteContainerMock.getTitle()).thenReturn("");
    when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mEditNoteContainerMock).setTitleError(any(String.class));
  }

  @Test
  public void onSaveNoteClickedWithEmptyContents_showsErrorMessage() {
    /* Given */
    mEditNotePresenter.setContainer(mEditNoteContainerMock);

    when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
    when(mEditNoteContainerMock.getContents()).thenReturn("");

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mEditNoteContainerMock).setContentsError(any(String.class));
  }

  @Test
  public void onSaveNoteClicked_persistsTheNote() {
    /* Given */
    mEditNotePresenter.setContainer(mEditNoteContainerMock);

    when(mEditNoteContainerMock.getTitle()).thenReturn(TITLE);
    when(mEditNoteContainerMock.getContents()).thenReturn(CONTENTS);

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mNoteRepository).update(mNote);
  }

  @Test
  public void onSaveNoteClicked_navigatesBack() {
    /* Given */
    EditNoteContainer containerMock = mock(EditNoteContainer.class);
    mEditNotePresenter.setContainer(containerMock);

    when(containerMock.getTitle()).thenReturn(TITLE);
    when(containerMock.getContents()).thenReturn(CONTENTS);

    /* When */
    mEditNotePresenter.onSaveNoteClicked();

    /* Then */
    verify(mTriad).goBack();
  }
}

File: triad-sample/app-kotlin/src/test/kotlin/com/nhaarman/triad/sample/editnote/EditExistingNotePresenterTest.kt
/*
 * Copyright 2016 Niek Haarman
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.nhaarman.triad.sample.editnote

import android.content.Context
import com.nhaarman.mockito.*
import com.nhaarman.triad.Triad
import com.nhaarman.triad.sample.*
import org.junit.Before
import org.junit.Test

class EditExistingNotePresenterTest {

    private lateinit var mEditNotePresenter: EditNotePresenter

    private lateinit var mEditNoteContainerMock: EditNoteContainer

    private lateinit var mNote: Note

    private lateinit var mNoteRepository: NoteRepository

    private lateinit var mNoteValidator: NoteValidator

    private lateinit var mTriad: Triad

    @Before
    fun setUp() {
        mNote = Note()
        mNote.title = TITLE
        mNote.contents = CONTENTS

        mNoteValidator = spy(NoteValidator())
        mNoteRepository = spy(MemoryNoteRepository())
        mTriad = mock()

        mEditNotePresenter = EditNotePresenter(mNote, mNoteValidator, mock(), mNoteRepository, mTriad)

        mEditNoteContainerMock = mock<EditNoteContainer>()
        whenever(mEditNoteContainerMock.context()).thenReturn(mock<Context>())
    }

    @Test
    fun onControlGained_setsNoteTitleToEditNoteContainer() {
        /* Given */
        val containerMock = mock<EditNoteContainer>()
        val activityComponentMock = mock<ActivityComponent>()

        /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock)

        /* Then */
        verify(containerMock).title = TITLE
    }

    @Test
    fun onControlGained_setsNoteContentsToEditNoteContainer() {
        /* Given */
        val containerMock = mock<EditNoteContainer>()
        val activityComponentMock = mock<ActivityComponent>()

        /* When */
        mEditNotePresenter.onControlGained(containerMock, activityComponentMock)

        /* Then */
        verify(containerMock).contents = CONTENTS
    }

    @Test
    fun onSaveNoteClicked_validatesTitle() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteValidator>(mNoteValidator).validateTitle(TITLE)
    }

    @Test
    fun onSaveNoteClicked_validatesContents() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteValidator>(mNoteValidator).validateContents(CONTENTS)
    }

    @Test
    fun onSaveNoteClickedWithEmptyTitle_showsErrorMessage() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn("")
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify(mEditNoteContainerMock).setTitleError(any())
    }

    @Test
    fun onSaveNoteClickedWithEmptyContents_showsErrorMessage() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn("")

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<EditNoteContainer>(mEditNoteContainerMock).setContentsError(any())
    }

    @Test
    fun onSaveNoteClicked_persistsTheNote() {
        /* Given */
        mEditNotePresenter.container = mEditNoteContainerMock

        whenever(mEditNoteContainerMock.title).thenReturn(TITLE)
        whenever(mEditNoteContainerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<NoteRepository>(mNoteRepository).update(mNote)
    }

    @Test
    fun onSaveNoteClicked_navigatesBack() {
        /* Given */
        val containerMock = mock<EditNoteContainer>()
        mEditNotePresenter.container = containerMock

        whenever(containerMock.title).thenReturn(TITLE)
        whenever(containerMock.contents).thenReturn(CONTENTS)

        /* When */
        mEditNotePresenter.onSaveNoteClicked()

        /* Then */
        verify<Triad>(mTriad).goBack()
    }

    companion object {

        private val TITLE = "title"

        private val CONTENTS = "contents"
    }
}